scanpipe/cyclonedx/__init__.py

```diff
@@ -0,0 +1,181 @@
+    """
+    if not root_component_list:
+        return
+
+    for component in root_component_list:
+        extra_data = (
+            bom_attributes_to_dict(component.components)
+            if component.components is not None
+            else {}
+        )
+        collected.append({"cdx_package": component, "nested_components": extra_data})
+        recursive_component_collector(component.components, collected)
+    return collected
+
+
+def resolve_license(license):
+    """
+    Return license expression/id/name from license item.
+    """
+    if "expression" in license:
+        return license["expression"]
+    elif "id" in license["license"]:
+        return license["license"]["id"]
+    else:
+        return license["license"]["name"]
+
+
+def get_declared_licenses(licenses):
+    """
+    Return resolved license from list of LicenseChoice.
+    """
+    if not licenses:
+        return ""
+
+    return "\n".join(
+        [resolve_license(license) for license in bom_attributes_to_dict(licenses)]
+    )
+
+
+def get_checksums(component: Component):
+    """
+    Return dict of all the checksums from a component.
+    """
+    if not component.hashes:
+        return {}
+
+    algorithm_map_cdx_scio = {
+        "MD5": "md5",
+        "SHA-1": "sha1",
+        "SHA-256": "sha256",
+        "SHA-512": "sha512",
+    }
+    return {
+        algorithm_map_cdx_scio[algo_hash.alg.value]: algo_hash.content.__root__
+        for algo_hash in component.hashes
+        if algo_hash.alg.value in algorithm_map_cdx_scio
+    }
+
+
+def get_external_refrences(external_references):
+    """
+    Return dict of refrence urls from list of `externalRefrences`.
+    """
+    if not external_references:
+        return {}
+
+    refrences = {
+        "vcs": [],
+        "issue-tracker": [],
+        "website": [],
+        "advisories": [],
+        "bom": [],
+        "mailing-list": [],
+        "social": [],
+        "chat": [],
+        "documentation": [],
+        "support": [],
+        "distribution": [],
+        "license": [],
+        "build-meta": [],
+        "build-system": [],
+        "release-notes": [],
+        "other": [],
+    }
+    for ref in external_references:
+        refrences[ref.type.value].append(ref.url)
+
+    return {key: value for key, value in refrences.items() if value}
+
+
+def validate_document(document, schema=CYCLONEDX_JSON_SCHEMA_PATH):
+    """
+    CYCLONEDX document validation.
+    """
+    if isinstance(document, str):
+        document = json.loads(document)
+
+    if isinstance(schema, pathlib.Path):
+        schema = schema.read_text()
+    if isinstance(schema, str):
+        schema = json.loads(schema)
+
+    resolver = jsonschema.RefResolver(
+        base_uri="file://" + str(pathlib.Path(__file__).parent), referrer=schema
+    )
+
+    validator = jsonschema.Draft7Validator(schema=schema, resolver=resolver)
+
+    validator.validate(instance=document)
```

Comment:
Is this complexity needed?
The basic `jsonschema.validate(instance=document, schema=schema)` is not enough in the CycloneDX case?

Reply:
`jsonschema.validate(instance=document, schema=schema)` will work, but if BOM contains an invalid SPDX license expression, then `hoppr-cyclonedx-models` won't be able to parse it. The same thing goes for JSF.